core/factorization/lu.cpp

```diff
@@ -0,0 +1,155 @@
+
+#include <ginkgo/core/factorization/lu.hpp>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/types.hpp>
+
+
+#include "core/components/fill_array_kernels.hpp"
+#include "core/factorization/lu_kernels.hpp"
+#include "core/factorization/symbolic.hpp"
+#include "core/matrix/csr_kernels.hpp"
+#include "core/matrix/csr_lookup.hpp"
+
+
+namespace gko {
+namespace experimental {
+namespace factorization {
+namespace {
+
+
+GKO_REGISTER_OPERATION(fill_array, components::fill_array);
+GKO_REGISTER_OPERATION(build_lookup_offsets, csr::build_lookup_offsets);
+GKO_REGISTER_OPERATION(build_lookup, csr::build_lookup);
+GKO_REGISTER_OPERATION(initialize, lu_factorization::initialize);
+GKO_REGISTER_OPERATION(factorize, lu_factorization::factorize);
+
+
+}  // namespace
+
+
+template <typename ValueType, typename IndexType>
+Lu<ValueType, IndexType>::Lu(std::shared_ptr<const Executor> exec,
+                             const parameters_type& params)
+    : EnablePolymorphicObject<Lu, LinOpFactory>(std::move(exec)),
+      parameters_(params)
+{}
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<Factorization<ValueType, IndexType>>
+Lu<ValueType, IndexType>::generate(
+    std::shared_ptr<const LinOp> system_matrix) const
+{
+    auto product =
+        std::unique_ptr<factorization_type>(static_cast<factorization_type*>(
+            this->LinOpFactory::generate(std::move(system_matrix)).release()));
+    return product;
+}
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<LinOp> Lu<ValueType, IndexType>::generate_impl(
+    std::shared_ptr<const LinOp> system_matrix) const
+{
+    GKO_ASSERT_IS_SQUARE_MATRIX(system_matrix);
+    const auto exec = this->get_executor();
+    // TODO deal with non Csr matrices
+    const auto mtx = as<matrix_type>(system_matrix);
+    const auto num_rows = mtx->get_size()[0];
+    std::unique_ptr<matrix_type> factors;
+    if (!parameters_.symbolic_factorization) {
+        if (parameters_.symmetric_sparsity) {
+            factors = gko::factorization::symbolic_cholesky(mtx.get());
+        } else {
+            GKO_NOT_SUPPORTED(mtx);
+        }
+    } else {
+        const auto& symbolic = parameters_.symbolic_factorization;
+        const auto factor_nnz = symbolic->get_num_nonzeros();
+        factors = matrix_type::create(exec, mtx->get_size(), factor_nnz);
+        const auto symbolic_exec = symbolic->get_executor();
+        exec->copy_from(symbolic_exec.get(), factor_nnz,
+                        symbolic->get_const_col_idxs(),
+                        factors->get_col_idxs());
+        exec->copy_from(symbolic_exec.get(), num_rows + 1,
+                        symbolic->get_const_row_ptrs(),
+                        factors->get_row_ptrs());
+        // update srow to be safe
+        factors->set_strategy(factors->get_strategy());
```

Comment:
we changed the row pointers of the matrix, so to avoid creating a partially valid matrix, we also need to update the strategy information based on the row pointers.